timemachine/fe/single_topology.py

```diff
@@ -596,17 +608,31 @@ def find_dummy_groups_and_anchors(
     solution later on is to minimize the number of rotatable bonds?
     """
 
-    assignments = generate_anchored_dummy_group_assignments(mol_a, mol_b, core_atoms_a, core_atoms_b)
+    bond_graph_a = convert_to_nx(mol_a)
+    bond_graph_b = convert_to_nx(mol_b)
+
+    atom_map_a_to_b = {a: b for a, b in zip(core_atoms_a, core_atoms_b)}
+    core_disabled_bonds = compute_disabled_bonds_in_core(bond_graph_a, bond_graph_b, core_atoms_a, atom_map_a_to_b)
```

Comment:
Two bugs here:

1. I think this returns bonds with indices are indexed w.r.t. `mol_a`'s atoms. But `dga_disabled_bonds = compute_disabled_bonds_in_dga(bond_graph_b, core_atoms_b, dgs)` generates bonds whose indices are indexed w.r.t. `mol_b`. Forming a union over these two sets is inconsistent.

2. Should run the `is_chiral_conversion` check on just the core bonds (in the event of there being no dummy-atoms).

Here's a test case:

```
def test_chiral_core_bond_breaking_raises_error():
    r"""
    Test that we raise assertions for molecules that cannot generate valid
    dummy-group anchor assignments. In particular, we break two core-core
    bonds under an identity mapping (with no dummy atoms).
    """
    mol_a = Chem.MolFromMolBlock("""
  Mrv2311 07312412093D          

  6  7  0  0  0  0            999 V2000
    1.7504    0.2003    1.2663 N   0  0  2  0  0  0  0  0  0  0  0  0
    0.8845   -0.7367    0.7929 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.0656    0.1573    0.2634 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.6778    2.1410    1.4645 F   0  0  0  0  0  0  0  0  0  0  0  0
    0.9965    1.1853    0.5313 C   0  0  1  0  0  0  0  0  0  0  0  0
    2.3507    0.8788    0.2466 O   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  5  4  1  0  0  0  0
  5  1  1  0  0  0  0
  2  3  1  0  0  0  0
  6  1  1  0  0  0  0
  3  5  1  0  0  0  0
  5  6  1  0  0  0  0
M  END
$$$$""", removeHs=False)
    
    mol_b = Chem.MolFromMolBlock("""
  Mrv2311 07312412083D          

  6  5  0  0  0  0            999 V2000
    1.7504    0.2003    1.2663 N   0  0  2  0  0  0  0  0  0  0  0  0
    0.8845   -0.7367    0.7929 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.0656    0.1573    0.2634 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.6778    2.1410    1.4645 F   0  0  0  0  0  0  0  0  0  0  0  0
    0.9965    1.1853    0.5313 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.3507    0.8788    0.2466 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  5  4  1  0  0  0  0
  5  1  1  0  0  0  0
  6  1  1  0  0  0  0
  2  3  1  0  0  0  0
M  END
$$$$""", removeHs=False)

    # need to generate SC charges on this mol - am1 fails
    ff = Forcefield.load_from_file('smirnoff_1_1_0_sc.py')
    core = np.array([[0,0],[1,1],[2,2],[3,3],[4,4],[5,5]])

    with pytest.raises(DummyGroupAssignmentError):
        SingleTopology(mol_a, mol_b, core, ff)
```

Reply:
note: I think the right thing to do here is to always return indices in terms of `mol_b` since later on we do:

```all_dummy_bond_idxs = recursive_map(all_dummy_bond_idxs, b_to_c)```

but please double check?

Reply:
Here's one more test (fails on master, fails on the branch as of commit [1b70e84]) that will pass if you modify:

``` python
    atom_map_b_to_a = {b: a for b, a in zip(core_atoms_b, core_atoms_a)}
    core_disabled_bonds = compute_disabled_bonds_in_core(bond_graph_b, bond_graph_a, core_atoms_b, atom_map_b_to_a)
```

``` python
def test_chiral_bond_breaking_1_core_1_dummy_v2():

    mol_a = Chem.MolFromMolBlock(
        """lhs
                    3D
 Structure written by MMmdl.
 15 16  0  0  1  0            999 V2000
    2.1455    1.6402   -0.1409 O   0  0  0  0  0  0
   -0.0529   -0.7121    0.3721 C   0  0  0  0  0  0
   -1.4286   -0.4301   -0.2680 C   0  0  0  0  0  0
   -0.9221    0.8962   -0.8723 C   0  0  0  0  0  0
    1.7163    0.6542   -1.0827 C   0  0  0  0  0  0
    0.4519    0.5990   -0.2512 C   0  0  0  0  0  0
    1.0309    1.5756    0.6040 O   0  0  0  0  0  0
   -0.0750   -0.6889    1.4650 H   0  0  0  0  0  0
    0.4285   -1.6073   -0.0287 H   0  0  0  0  0  0
   -2.2238   -0.2880    0.4685 H   0  0  0  0  0  0
   -1.6971   -1.1659   -1.0311 H   0  0  0  0  0  0
   -1.4047    1.7800   -0.4452 H   0  0  0  0  0  0
   -0.9239    0.8937   -1.9651 H   0  0  0  0  0  0
    1.5863    1.0639   -2.0863 H   0  0  0  0  0  0
    2.3266   -0.2504   -1.0468 H   0  0  0  0  0  0
  6  2  1  0  0  0
  6  4  1  0  0  0
  6  5  1  0  0  0
  6  7  1  0  0  0
  2  3  1  0  0  0
  2  8  1  0  0  0
  2  9  1  0  0  0
  3  4  1  0  0  0
  3 10  1  0  0  0
  3 11  1  0  0  0
  4 12  1  0  0  0
  4 13  1  0  0  0
  5  1  1  0  0  0
  5 14  1  0  0  0
  5 15  1  0  0  0
  1  7  1  0  0  0
M  END""",
        removeHs=False,
    )

    mol_b = Chem.MolFromMolBlock(
        """rhs
                    3D
 Structure written by MMmdl.
 12 11  0  0  1  0            999 V2000
    0.4102    0.4907   -0.1997 O   0  0  0  0  0  0
   -0.0529   -0.7121    0.3721 C   0  0  0  0  0  0
   -1.4286   -0.4301   -0.2680 C   0  0  0  0  0  0
   -1.0708    0.5068   -0.6949 H   0  0  0  0  0  0
    1.5875    0.5421   -0.9739 C   0  0  0  0  0  0
    1.9148    1.2939   -0.2558 H   0  0  0  0  0  0
   -0.0750   -0.6889    1.4650 H   0  0  0  0  0  0
    0.4285   -1.6073   -0.0287 H   0  0  0  0  0  0
   -2.2238   -0.2880    0.4685 H   0  0  0  0  0  0
   -1.6971   -1.1659   -1.0311 H   0  0  0  0  0  0
    1.4575    0.9518   -1.9775 H   0  0  0  0  0  0
    2.1978   -0.3625   -0.9380 H   0  0  0  0  0  0
  1  2  1  0  0  0
  1  5  1  0  0  0
  2  3  1  0  0  0
  2  7  1  0  0  0
  2  8  1  0  0  0
  3  4  1  0  0  0
  3  9  1  0  0  0
  3 10  1  0  0  0
  5  6  1  0  0  0
  5 11  1  0  0  0
  5 12  1  0  0  0
M  END
""",
        removeHs=False,
    )

    # need to generate SC charges on this mol - am1 fails
    ff = Forcefield.load_from_file("smirnoff_1_1_0_sc.py")
    core = np.array(
        [[5, 0], [1, 1], [2, 2], [3, 3], [4, 4], [0, 5], [7, 6], [8, 7], [9, 8], [10, 9], [13, 10], [14, 11]]
    )

    from timemachine.fe.utils import plot_atom_mapping_grid

    print("core", core)
    res = plot_atom_mapping_grid(mol_a, mol_b, core)
    fpath = f"atom_mapping.svg"
    print("core mapping written to", fpath)

    with open(fpath, "w") as fh:
        fh.write(res)

    # should not raise an assertion
    SingleTopology(mol_a, mol_b, core, ff)
```

Reply:
Added your test in https://github.com/proteneer/timemachine/pull/1356/commits/8b66a3524271c30d3e9d1a0d76e709dadc2c7171 and fixed in https://github.com/proteneer/timemachine/pull/1356/commits/2a8f88f22a660590582498bccd2727f1b7e98743

Reply:
Added your first test in https://github.com/proteneer/timemachine/pull/1356/commits/8b66a3524271c30d3e9d1a0d76e709dadc2c7171 and fixed in https://github.com/proteneer/timemachine/pull/1356/commits/2a8f88f22a660590582498bccd2727f1b7e98743. Will now look into the second test.

Reply:
Added second test case in https://github.com/proteneer/timemachine/pull/1356/commits/9dc0885bff6716c61a0332861c61dd525dfbcaf2. (It passes since https://github.com/proteneer/timemachine/pull/1356/commits/2a8f88f22a660590582498bccd2727f1b7e98743)

Reply:
> note: I think the right thing to do here is to always return indices in terms of `mol_b` since later on we do:
> 
> `all_dummy_bond_idxs = recursive_map(all_dummy_bond_idxs, b_to_c)`
> 
> but please double check?

Correct, our convention is that indices in the dummy group assignments refer to atoms in mol_b. Tried to clarify in a few places in https://github.com/proteneer/timemachine/pull/1356/commits/c1e9d3ce91d5c465ae27b025fe5430486c8e4907